lightly/loss/barlow_twins_loss.py

```diff
@@ -0,0 +1,47 @@
+import torch
+
+class BarlowTwinsLoss(torch.nn.Module):
+    """Implementation of the Barlow Twins Loss from Barlow Twins[0] paper.
+    This code specifically implements the Figure Algorithm 1 from [0].
+    [0] Zbontar,J. et.al, 2021, Barlow Twins... https://arxiv.org/abs/2103.03230
+
+        Examples:
+
+        >>> # initialize loss function
+        >>> loss_fn = BarlowTwinsLoss()
+        >>>
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # feed through SimSiam model
+        >>> out0, out1 = model(t0, t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+
+    """
+
+    def __init__(self, device, lambda_param=5e-3):
+        super(BarlowTwinsLoss, self).__init__()
+        self.lambda_param = lambda_param
+        self.device = device
+
+    def forward(self, z_a: torch.Tensor, z_b: torch.Tensor):
+        # normalize repr. along the batch dimension
+        z_a_norm = (z_a - z_a.mean(0)) / z_a.std(0) # NxD
+        z_b_norm = (z_b - z_b.mean(0)) / z_b.std(0) # NxD
+
+        N = z_a.size(0)
+        D = z_a.size(1)
+
+        # cross-correlation matrix
+        c = torch.mm(z_a_norm.T, z_b_norm) / N # DxD
+        # loss
+        c_diff = (c - torch.eye(D, device=self.device)).pow(2) # DxD
```

Comment:
When you change the way you get the device, don't forget to change it here as well :)

Reply:
Thanks :)